File: android/app/src/main/kotlin/dev/agentdeck/terrarium/TerrariumState.kt
package dev.agentdeck.terrarium

import android.util.Log
import dev.agentdeck.net.AgentState
import dev.agentdeck.state.DashboardState
import dev.agentdeck.terrarium.creature.AgentMark

private const val TAG = "Terrarium"

/** Visual states for each creature and the environment. */

enum class OctopusVisualState {
    SLEEPING,    // Curled up at bottom, dim, eyes closed
    FLOATING,    // Gentle sine bob, tentacles wave
    WORKING,     // Starburst animation — processing (tool use or thinking)
    ASKING,      // Speech bubble + "?" — awaiting user input
}

enum class CrayfishVisualState {
    DORMANT,    // Partially hidden behind rocks
    SITTING,    // Idle on rock, claws at rest
    OBSERVING,  // Watching activity — gentle claw fidget, eyes tracking
    ROUTING,    // Claws clap, eyes flash, signal lines emit (OpenClaw orchestrating)
    WAITING,    // Claws raised
}

enum class TetraVisualState {
    ABSENT,     // Not visible (disconnected)
    CIRCLING,   // Boids algorithm, orbiting attractor
    STREAMING,  // Line up, streak horizontally, code trail particles
    HOVERING,   // Formation near options area
}

enum class EnvironmentVisualState {
    DARK,    // Disconnected — dim/off
    CALM,    // Idle — gentle caustics, slow bubbles
    ACTIVE,  // Processing — bright caustics, more bubbles
    ALERT,   // Awaiting input — pulsing highlights
}

/** Per-agent creature state for multi-session rendering. */
data class AgentCreatureState(
    val sessionId: String,
    val agentType: String?,
    val mark: AgentMark?,
    val visualState: OctopusVisualState,
    val isPrimary: Boolean,
    val layoutSlot: Int,
    val displayName: String? = null,
)

/** Combined visual state for the entire terrarium scene. */
data class TerrariumState(
    val octopus: OctopusVisualState,
    val crayfish: CrayfishVisualState,
    val tetra: TetraVisualState,
    val environment: EnvironmentVisualState,
    val currentTool: String? = null,
    val toolProgress: String? = null,
    val projectName: String? = null,
    val modelName: String? = null,
    val agentType: String? = null,
    val hasError: Boolean = false,
    /** Multi-session: all coding agent creatures (octopuses). */
    val agents: List<AgentCreatureState> = emptyList(),
    /** OpenClaw backend worker count. */
    val workerCrayfishCount: Int = 0,
)

/** Map DashboardState to visual TerrariumState. */
fun DashboardState.toTerrariumState(): TerrariumState {
    val isOpenClaw = agentType == "openclaw"
    val hasTool = currentTool != null
    Log.d(TAG, "toTerrariumState: agentState=$agentState, agentType=$agentType, isOpenClaw=$isOpenClaw, hasTool=$hasTool")

    val octopus = when (agentState) {
        AgentState.DISCONNECTED -> OctopusVisualState.SLEEPING
        AgentState.IDLE -> OctopusVisualState.FLOATING
        AgentState.PROCESSING -> OctopusVisualState.WORKING
        AgentState.AWAITING_PERMISSION,
        AgentState.AWAITING_OPTION,
        AgentState.AWAITING_DIFF -> OctopusVisualState.ASKING
    }

    // OpenClaw sibling state determines crayfish independently
    val ocSibling = siblingSessions.firstOrNull { it.agentType == "openclaw" }
    val crayfish = when {
        // Primary is OpenClaw — use primary state
        isOpenClaw -> when (agentState) {
            AgentState.PROCESSING -> CrayfishVisualState.ROUTING
            AgentState.IDLE -> CrayfishVisualState.SITTING
            AgentState.DISCONNECTED -> CrayfishVisualState.DORMANT
            else -> CrayfishVisualState.WAITING
        }
        // Sibling OpenClaw exists — use its state
        ocSibling != null -> when (ocSibling.state) {
            "processing" -> CrayfishVisualState.ROUTING
            "idle" -> CrayfishVisualState.SITTING
            "awaiting_permission", "awaiting_option", "awaiting_diff" -> CrayfishVisualState.WAITING
            else -> if (ocSibling.alive) CrayfishVisualState.SITTING else CrayfishVisualState.DORMANT
        }
        // Gateway detected but no bridge
        gatewayAvailable == true -> CrayfishVisualState.SITTING
        // Nothing — derive from primary agent state
        else -> when (agentState) {
            AgentState.DISCONNECTED -> CrayfishVisualState.DORMANT
            AgentState.PROCESSING -> CrayfishVisualState.OBSERVING
            else -> CrayfishVisualState.SITTING
        }
    }

    val tetra = when (agentState) {
        AgentState.DISCONNECTED -> TetraVisualState.ABSENT
        AgentState.IDLE -> TetraVisualState.CIRCLING
        AgentState.PROCESSING -> if (hasTool) TetraVisualState.STREAMING else TetraVisualState.CIRCLING
        AgentState.AWAITING_PERMISSION,
        AgentState.AWAITING_OPTION,
        AgentState.AWAITING_DIFF -> TetraVisualState.HOVERING
    }

    val environment = when (agentState) {
        AgentState.DISCONNECTED -> EnvironmentVisualState.DARK
        AgentState.IDLE -> EnvironmentVisualState.CALM
        AgentState.PROCESSING -> EnvironmentVisualState.ACTIVE
        AgentState.AWAITING_PERMISSION,
        AgentState.AWAITING_OPTION,
        AgentState.AWAITING_DIFF -> EnvironmentVisualState.ALERT
    }

    Log.d(TAG, "Terrarium mapped: octopus=$octopus, crayfish=$crayfish, tetra=$tetra, env=$environment")

    // Build multi-agent creature list from sibling sessions
    val agents = mutableListOf<AgentCreatureState>()

    // Primary agent (currently connected session)
    agents.add(
        AgentCreatureState(
            sessionId = sessionId ?: "primary",
            agentType = agentType,
            mark = AgentMark.fromAgentType(agentType),
            visualState = octopus,
            isPrimary = true,
            layoutSlot = 0,
            displayName = projectName,
        )
    )

    // Sibling sessions (coding agents only — not the current session)
    var slot = 1
    for (sibling in siblingSessions) {
        if (sessionId != null && sibling.id == sessionId) continue // skip self (null guard)
        val siblingType = sibling.agentType
        if (siblingType == "openclaw") continue // crayfish, not octopus
        agents.add(
            AgentCreatureState(
                sessionId = sibling.id,
                agentType = siblingType,
                mark = AgentMark.fromAgentType(siblingType),
                visualState = mapSessionOctopusState(sibling.state),
                isPrimary = false,
                layoutSlot = slot++,
                displayName = sibling.projectName,
            )
        )
    }

    return TerrariumState(
        octopus = octopus,
        crayfish = crayfish,
        tetra = tetra,
        environment = environment,
        currentTool = currentTool,
        toolProgress = toolProgress,
        projectName = projectName,
        modelName = modelName,
        agentType = agentType,
        agents = agents,
        workerCrayfishCount = workerSessionCount ?: 0,
    )
}

private fun mapSessionOctopusState(state: String?): OctopusVisualState = when (state) {
    "processing" -> OctopusVisualState.WORKING
    "awaiting_permission", "awaiting_option", "awaiting_diff" -> OctopusVisualState.ASKING
    "idle" -> OctopusVisualState.FLOATING
    else -> OctopusVisualState.FLOATING
}

File: apple/AgentDeck/Terrarium/TerrariumState.swift
// TerrariumState.swift — Visual state mapping from DashboardState
// Ported from android TerrariumState.kt

import Foundation

// MARK: - Octopus Visual State

enum OctopusVisualState {
    case sleeping   // Curled at bottom, dim
    case floating   // Gentle bob, tentacle wave (idle)
    case working    // Swim + starburst (processing)
    case asking     // Fidget + "?" bubble (awaiting input)
}

// MARK: - Crayfish Visual State

enum CrayfishVisualState {
    case dormant    // Hidden behind rocks
    case sitting    // Idle, heartbeat glow
    case observing  // Watching, gentle fidget
    case routing    // Claws clap, signal waves
    case waiting    // Claws raised
    case sick       // Desaturated, tilted (gateway error)
}

// MARK: - Tetra Visual State

enum TetraVisualState {
    case absent
    case circling   // Boids orbit
    case streaming  // Line up, food chase
    case hovering   // Near options
}

// MARK: - Agent Creature State

struct AgentCreatureState: Identifiable {
    let id: String           // session ID
    let projectName: String?
    let modelName: String?
    let state: OctopusVisualState
    let homeX: Float
    let homeY: Float
    let scale: Float

    /// Whether this session just exited ASKING (triggers pop burst)
    var exitedAsking = false
}

// MARK: - Terrarium State (aggregate)

struct TerrariumState {
    var creatures: [AgentCreatureState] = []
    var crayfishState: CrayfishVisualState = .dormant
    var crayfishVisible: Bool = false
    var tetraState: TetraVisualState = .circling
    var environment: EnvironmentVisualState = .calm
    var hasError: Bool = false

    /// Pop burst positions (from ASKING exit)
    var popBurstPositions: [(x: Float, y: Float)] = []
}

// MARK: - Mapping from DashboardState

extension DashboardState {
    func toTerrariumState(previous: TerrariumState? = nil) -> TerrariumState {
        var result = TerrariumState()

        // Primary session creature (skip daemon and openclaw — they're not octopuses)
        let primaryIsOctopus = state != .disconnected
            && agentType != "daemon"
            && agentType != "openclaw"

        // Octopus siblings (exclude daemon + openclaw)
        let siblings = siblingSessions.filter {
            $0.agentType != "daemon" && $0.agentType != "openclaw"
        }

        let octopusCount = (primaryIsOctopus ? 1 : 0) + siblings.count
        let slots = CreatureLayout.layoutOctopuses(count: max(1, octopusCount))

        var creatures: [AgentCreatureState] = []
        var slotIdx = 0

        if primaryIsOctopus {
            let primaryState = mapToOctopusState(state)
            let slot = slots.first ?? CreatureSlot(x: 0.4, y: 0.45, scale: 1.0)

            // Detect ASKING exit for pop burst
            let wasAsking = previous?.creatures.first?.state == .asking
            let exitedAsking = wasAsking && primaryState != .asking

            creatures.append(AgentCreatureState(
                id: sessionId ?? "primary",
                projectName: projectName,
                modelName: modelName,
                state: primaryState,
                homeX: slot.x,
                homeY: slot.y,
                scale: slot.scale,
                exitedAsking: exitedAsking
            ))
            slotIdx = 1
        }
        // Sibling octopuses
        for (i, sibling) in siblings.enumerated() {
            let idx = min(slotIdx + i, slots.count - 1)
            let s = slots[idx]
            let sibState = mapSiblingState(sibling.state)
            creatures.append(AgentCreatureState(
                id: sibling.id,
                projectName: sibling.projectName,
                modelName: nil,
                state: sibState,
                homeX: s.x,
                homeY: s.y,
                scale: s.scale
            ))
        }

        result.creatures = creatures

        // Environment state
        result.environment = mapToEnvironment(state)
        result.hasError = gatewayHasError

        // Crayfish (OpenClaw gateway)
        result.crayfishVisible = gatewayAvailable

        if gatewayHasError {
            result.crayfishState = .sick
        } else if let ocSibling = siblingSessions.first(where: { $0.agentType == "openclaw" }) {
            result.crayfishState = ocSibling.state == "processing" ? .routing : .sitting
        } else if gatewayAvailable {
            result.crayfishState = .sitting
        } else {
            result.crayfishState = .dormant
        }

        // Tetra state
        if state == .processing || creatures.contains(where: { $0.state == .working }) {
            result.tetraState = .streaming
        } else if state.isAwaiting {
            result.tetraState = .hovering
        } else {
            result.tetraState = .circling
        }

        // Pop bursts from ASKING exits
        result.popBurstPositions = creatures
            .filter { $0.exitedAsking }
            .map { (x: $0.homeX, y: $0.homeY) }

        return result
    }

    private func mapToOctopusState(_ connState: AgentConnectionState) -> OctopusVisualState {
        switch connState {
        case .disconnected: .sleeping
        case .idle: .floating
        case .processing: .working
        case .awaitingPermission, .awaitingOption, .awaitingDiff: .asking
        }
    }

    private func mapSiblingState(_ stateStr: String?) -> OctopusVisualState {
        switch stateStr {
        case "processing": .working
        case "awaiting_permission", "awaiting_option", "awaiting_diff": .asking
        case "idle": .floating
        default: .sleeping
        }
    }

    private func mapToEnvironment(_ connState: AgentConnectionState) -> EnvironmentVisualState {
        switch connState {
        case .disconnected: .dark
        case .idle: .calm
        case .processing: .active
        case .awaitingPermission, .awaitingOption, .awaitingDiff: .alert
        }
    }
}
